⚙️ add auto-mod like functionality

teauxfu opened the pull request:
Excuse my rusty Python, and my addition of type hints. I'm rather used to C# lately and it made things easier to read. I had to make a couple choices here and there let me know what you think 

- Opted to just write a few helpers to initialize the db on startup. Didn't move to include anything ORM-like, or anything to deal with migrations, etc. 
- Since the new behavior to delete the offending message does so, the logic in the emoji reaction block had to change a bit. I came to see that we were doing a kind of "auth check" to see if the user reacting was the original poster. Since that original message would be gone already, it would throw an error. My workaround to this was to add another small table to log messages and assoc. user id of those deleted by the bot. When handling an emoji reaction, we first check the db to see if that's one we already deleted. If we don't get a match, then we do the API call. We compact the db on app startup by deleting any entries older than 14 days.
- For the sake of not making the code look paranoid, I didn't add any kind of error handling really 

One part I haven't started on yet is how to allow server admins to actually set the setting. I think that may mean reaching for a slash command or something, and possibly also checking if that calling user has permission to update the setting. I suppose we'd want only some users to be able to change that. Let me know if you have any suggestions in that regard 

Edit: I've added the slash command `/linkautomod`  which takes one bool `enabled` to toggle the automod behavior. Let me know if you have any suggestions or concerns around renaming anything 

Overall it seems to be working as intended

teauxfu committed:
start modding, add db

teauxfu committed:
add to docs

teauxfu committed:
add database stuff

teauxfu committed:
implement the autodelete functionality

teauxfu committed:
cleanup

teauxfu committed:
fix gitignore

teauxfu committed:
cleanup

teauxfu committed:
implement a slash command to toggle the automod behavior

teauxfu committed:
cleanup

teauxfu committed:
update hosting instructions in readme

danielzting commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,49 @@
+
+`python -m venv .`
+
+Activate the environment with `scripts\activate` (Windows) or `bin\activate` (Linux).
```
macOS/Linux command should be `source bin/activate`

danielzting commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,49 @@
+
+## Project structure 
+
+The [unalix](https://github.com/AmanoTeam/Unalix/tree/master/unalix) library is archived. Its source is included in this project in the `unalix` folder.
```
Can you also mention that the data.min.json file is sourced from https://docs.clearurls.xyz/latest/specs/rules/ as the one provided by Unalix was outdated, and that pull requests to update this file to keep it in sync with the official ClearURLs ruleset are welcome

danielzting commented on database/__init__.py:
Is this file required?

teauxfu replied:
looks like it's not, thanks

danielzting commented on database/guildsettings.py:
```diff
@@ -0,0 +1,44 @@
+def ensure_settings_table(con: sqlite3.Connection) -> None:
+    sql = """
+    create table if not exists 
+    GuildSettings_DeleteAndRepostDirtyMessages (guild_id int primary key, is_enabled bit);
```
As a day Java dev I appreciate the descriptiveness—but can we change the table name to `guild_settings`, then `is_enabled` can be something like `replace_messages`. This way we can reduce verbosity while allowing the possibility of adding more settings as columns later while respecting the SQL convention of snake_case

While you're at it could you also capitalize keywords like CREATE TABLE IF NOT EXISTS ...

teauxfu replied:
Sure, that sounds fine. I was trying to keep it super simple -- what with the lack of migrations and all. The idea here was to keep the feature as isolated as possible, while still suggesting a general pattern to attach later features. 

On a personal level, I find the all caps keywords jarring, but I will change them

danielzting commented on database/guildsettings.py:
```diff
@@ -0,0 +1,44 @@
+    cur.execute(sql)
+    cur.close()
+
+def ensure_guild_automod_setting(con: sqlite3.Connection, guild_id: int) -> None:
```
Is this function needed? I see you already handle the case where the guild is not in the table using coalesce which is good design, so I'm not too sure what's the purpose of this function

teauxfu replied:
the intention here was just to have an way of ensuring a subsequent update call works,

danielzting commented on database/util.py:
```diff
@@ -0,0 +1,20 @@
+        ensure_deleted_messages_table(con)
+        compact_deleted_messages_table(con)
+        for guild in guilds:
+            ensure_guild_automod_setting(con, guild.id)
```
Is this necessary? This looks like it's gonna send O(N) queries. If there's 100 guilds we don't want to send 100 round trips to the database

teauxfu replied:
The current methods to change the setting do an update, so I think the idea at the time was to make sure the updates would always be safe. I wouldn't really sweat about running many small queries one time against a local database file at startup, but we can just do an existence check at update time instead.

danielzting commented on main.py:
```diff
@@ -20,26 +27,37 @@
-    await client.change_presence(activity=discord.Activity(type=discord.ActivityType.watching, name='for tracking links'))
+    with get_dbcon() as con:        
+        initialize_db(con, bot.guilds)
+    con.close()
```
Do you need this? Doesn't the with-as automatically close

teauxfu replied:
Apparently it does not automatically close. I also expected that it would. https://docs.python.org/3/library/sqlite3.html#how-to-use-the-connection-context-manager

danielzting replied:
Huh, interesting. Could you use [contextlib.closing](https://discuss.python.org/t/implicitly-close-sqlite3-connections-with-context-managers/33320/5) or [contextmanager](https://stackoverflow.com/a/67436763), whichever you prefer stylistically, so we don't have all these con.close calls cluttering the code and avoid the possibility of forgetting to close?

teauxfu replied:
I don't know that I'd call three calls clutter, but sure, I've added contextlib.closing.

danielzting commented on main.py:
```diff
@@ -20,26 +27,37 @@
+
+    # sync our registered bot commands to the server
+    await bot.tree.sync()
+    await bot.change_presence(activity=discord.Activity(type=discord.ActivityType.watching, name='for tracking links'))
```
Apparently it's best practice to set the activity in the constructor https://stackoverflow.com/a/72848078

danielzting commented on main.py:
```diff
@@ -20,26 +27,37 @@
+    con.close()
+
+    # sync our registered bot commands to the server
+    await bot.tree.sync()
```
Likewise calling sync in on_ready has been said to trigger rate limits, would it be possible to create a command decorated with is_owner that only the bot owner can use to sync?

teauxfu replied:
Good idea, after skimming the docs a bit I think we should consider doing it in https://discordpy.readthedocs.io/en/stable/api.html?highlight=on_ready#discord.Client.setup_hook

Looks like this way it'll only run once

danielzting replied:
Just to make sure, you never ran into rate limit errors using setup_hook, even when restarting multiple times in a short time period during development?

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+    """Enable or disable link automoderation for this server"""
     
+    # check if user is guild owner or bot owner
+    is_owner = await bot.is_owner(interaction.user)
```
Don't we only need to check for guild owner? If I join some arbitrary server that uses my hosted instance I shouldn't be allowed to change the setting for that server right

teauxfu replied:
hmm yeah, fair point. I think the idea at the time was the bot owner presumably has read/write db access, so it seemed the same result. I'll remove this.

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+        
+    except Exception as e:
+        await interaction.response.send_message(f"❌ An error occurred: {str(e)}", ephemeral=True)
+        logger.error(f"Error in linkautomod command: {e}", e)
```
Does this print the full stack trace? According to my research logger.exception is the idiomatic way to log exceptions including the message and stack trace

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+        await interaction.response.send_message(f"✅ Link automod has been **{status}** for this server.")
+        
+    except Exception as e:
+        await interaction.response.send_message(f"❌ An error occurred: {str(e)}", ephemeral=True)
```
Could you also add a note linking this repo's issues page and informing the user they may report an issue here?

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+        with get_dbcon() as con:
+            ensure_guild_automod_setting(con, interaction.guild.id)
+            set_automod_setting(con, interaction.guild.id, enabled)
+        con.close()
```
unnecessary?

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
         deleted_messages.inc()
+        logger.info("Deleted message %i via reaction by user %i", message.id, payload.user_id)
+
+@bot.tree.command(name="linkautomod", description="Enable or disable link automoderation")
```
Would it be possible to describe what automoderation means—at least I am not familiar with this term and would not be able to conclude that this means the bot would delete the original message just from this description

teauxfu replied:
That's a fair point. Honestly I was just grabbing for some words to describe the feature enough to code it, I was hoping you'd have some suggestions around naming / what to call it. 

I am familiar with the term because in many Discord servers there's an automatic moderation feature that can filter certain kinds of content. I can appreciate that not everyone will be familiar with the expression.

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+        with get_dbcon() as con:
+            ensure_guild_automod_setting(con, message.guild.id)
+            should_delete_dirty_message = get_automod_setting(con, message.guild.id)
+        con.close()
```
ditto

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
-        # Send message and add reactions
-        if cleaned:
+        if not cleaned:
+            return
```
Nothing to change here, just wanted to say that with the increasing complexity of the cleaning logic early return here is good design 👍

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+            text = f'It appears that you have sent one or more links with tracking parameters. Below are the same links with those fields removed:\n{"\n".join(cleaned)}'
             await message.reply(text, mention_author=False)
             cleaned_messages.inc()
+            return 
```
instead of return here, can you wrap the rest of the function in an else block? I think that conveys the intent of your logic better

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+        await message.delete()
+        with get_dbcon() as con:
+            insert_deleted_message(con, message)
+        con.close()
```
"

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+            cleaned_content = cleaned_content.replace(url, clear_url(url))
+        text = f'User {message.author} sent the following message, which was deleted and has automatically been cleaned from tracking links:\n\n{cleaned_content}'
+        await message.reply(text, mention_author=False)
+        cleaned_messages.inc()
```
`cleaned_messages.inc` runs in both cases and can be factored out before the if statement

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+        for url in urls:
+            cleaned_content = cleaned_content.replace(url, clear_url(url))
+        text = f'User {message.author} sent the following message, which was deleted and has automatically been cleaned from tracking links:\n\n{cleaned_content}'
+        await message.reply(text, mention_author=False)
```
Out of curiosity what does replying do? Wouldn't it show "original message was deleted?"

teauxfu replied:
Yeah I believe it does. That seemed preferable at the time for context/clarity, but I can appreciate that one may consider it visual clutter. We can just send a new message instead if that's preferable

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+        cleaned_content = message.content
+        for url in urls:
+            cleaned_content = cleaned_content.replace(url, clear_url(url))
+        text = f'User {message.author} sent the following message, which was deleted and has automatically been cleaned from tracking links:\n\n{cleaned_content}'
```
can you add a "which was deleted and has automatically been cleaned from tracking links (what's this?):" where "what's this" links to a new section on the homepage that explains the setting and that a server admin enabled it?

danielzting replied:
BTW, the homepage is a static GitHub pages site in the `/docs` folder in case you missed it

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+    if permissions.manage_messages and original_message.author == reacting_user:
         await message.delete()
         deleted_messages.inc()
+        logger.info("Deleted message %i via reaction by user %i", message.id, payload.user_id)
```
In another place you use f-strings in logger, could you use either f-strings or % everywhere for consistency

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+    # before we check the API to do this "auth check", see if this is a message we deleted
+    con = get_dbcon()
+    user_id_whose_message_we_deleted = get_deleted_message_author(con, message.reference.message_id)
+    con.close()
```
can we use a with-as here like you did everywhere else

danielzting commented on main.py:
```diff
@@ -55,38 +73,99 @@ async def on_message(message):
+    reacting_user = await bot.fetch_user(payload.user_id)
+    # before we check the API to do this "auth check", see if this is a message we deleted
+    con = get_dbcon()
+    user_id_whose_message_we_deleted = get_deleted_message_author(con, message.reference.message_id)
```
I think `deleted_message_author` would be a more concise name, and mirrors your well-named function

danielzting commented on database/auditlog.py:
I think we can make deleting work without the database by checking the author mentioned in the bot's message. You already include it as part of "User {message.author} sent the following message..." Assuming `message.author` uniquely identifies the author, can you check if the user that clicked on the trashcan emoji is equal to the author mentioned in the bot's message? This approach you have looks like it would work fine, but not having to maintain another table and db connection would reduce complexity and help with future scalability

teauxfu replied:
Oh, good idea. I didn't realize the message object has a mentions property. So, yeah, we can just look for the user id of the mentioned user and compare that. Much cleaner, thanks.

danielzting commented on CONTRIBUTING.md:
Thanks for writing this up!

danielzting commented on README.md:
If you want, feel free to add a contributors section with yourself in it

danielzting reviewed (changes requested):
Was away for a couple days, thanks for the PR and your patience! Have some asks to improve code design. I only use Python as a hobby language so please let me know if I misunderstood something.

danielzting commented on README.md:
```diff
@@ -14,20 +14,26 @@ Disclaimer: This public instance is hosted on my Raspberry Pi and I make no guar
-4. Add a bot and check the `bot` scope and the above permissions in the OAuth2 tab
+4. Add a bot and check the `bot` and `applications.commands` scopes and the above permissions in the OAuth2 tab
 5. Visit the generated link to invite the bot
 6. Copy the token from the Bot tab and paste `TOKEN=[your clipboard here]` into a file named `.env`
```
Could you also mention the new DATABASE env var here too

teauxfu committed:
update docs

teauxfu committed:
remove junk

teauxfu committed:
simplify guild settings

teauxfu committed:
remove junk

teauxfu committed:
cleanup

teauxfu committed:
cleanup

teauxfu committed:
make sql keywords all caps

teauxfu committed:
make description <= 100 chars

teauxfu commented:
Thanks for the feedback, some good catches in here. I think this is ready for another review, I'm pretty sure I addressed all your points. 

edit: set this back to draft, still has a couple rough edges

teauxfu committed:
misc cleanup

teauxfu committed:
fix link

teauxfu committed:
update docs

teauxfu committed:
update doc

teauxfu committed:
cleanup link to what's this

teauxfu committed:
fix typo

teauxfu committed:
wip, the reaction handler is behaving inconsistent across channels

teauxfu commented:
Sorry, I came back with the edits a little too soon. I cleaned up some confused logic in the reaction handler, but now I'm getting weird behavior. The bot behaves correctly in all my test server channels except for one. I'll have to revisit this later as time allows

teauxfu committed:
early return on ephemeral messages

teauxfu commented:
Alrighty so I spun up a fresh Discord server, added the bot there, and things worked as expected. I'm not really sure what the problem was the other day, but maybe removing and reinstalling the bot would have helped. 

I had a couple friends join a test server to test the bot functions in either mode, and to test the `/linkautomod` command, and everything seemed to work as expected there as well. Can you please take another look @danielzting?

danielzting commented on README.md:
```diff
@@ -12,22 +12,42 @@ Have you ever noticed those `utm` and `fbclid` snippets in links? Those are used
+
+### Enabling automoderation
+
+Server admins can enable an automatic moderation feature where messages with dirty links are automatically deleted, then the bot sends a mesage with a claned version of the original message content. 
```
```suggestion
Server admins can enable an automatic moderation feature where messages with dirty links are automatically deleted, then the bot sends a mesage with a cleaned version of the original message content. 
```

danielzting commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,53 @@
+
+The bot uses a sqlite table to persist server settings. Functions to initialize and interact with the database are in the `database` folder.
+
+`guildsettings.py` is for Discord server specific settings. Currently we use a small lookup table to store that flag.
```
I think a first time contributor would be confused by "that flag" since this is the first time it's mentioned. TBH I don't think explaining what a database table is is necessary lol

danielzting commented on main.py:
```diff
@@ -55,38 +75,104 @@ async def on_message(message):
+            # Send message and add reactions
+            text = f"It appears that {message.author.mention} sent one or more links with tracking parameters. Below are the same links with with tracking parameters removed {whats_this}:\n\n{"\n".join(cleaned)}"
+            await message.reply(text, silent=True)
+        else :
```
```suggestion
        else:
```

danielzting commented on main.py:
```diff
@@ -55,38 +75,104 @@ async def on_message(message):
+        except discord.Forbidden:
+            logger.exception(f"Missing permissions to delete message {message.id} in channel {channel.name}")
+        except discord.HTTPException as e:
+            logger.exception(f"Failed to delete message {message.id}: {e}")
```
You don't need to pass in {e} right, logger.exception should include the full stack trace

danielzting commented on main.py:
```diff
@@ -55,38 +75,104 @@ async def on_message(message):
+async def set_automod_behavior(interaction: discord.Interaction, enabled: bool):
+    """Enable or disable link automoderation for this server"""
     
+    # check if user is guild owner or bot owner
```
```suggestion
    # check if user is guild owner
```

danielzting commented on main.py:
```diff
@@ -55,38 +75,104 @@ async def on_message(message):
+    # check if user is guild owner or bot owner
+    is_guild_owner = interaction.guild.owner_id == interaction.user.id
+    
+    if not (is_guild_owner):
```
parens unnecessary

Also, do you think restricting this to server owners is too strict? In my head if someone can add the bot they should be able to change the settings, i.e. has the manage server permission. I see you added `@commands.has_permissions(administrator=True)` which I think is fine by itself.

danielzting commented on main.py:
```diff
@@ -55,38 +75,104 @@ async def on_message(message):
+        
+    except Exception as e:
+        await interaction.response.send_message(f"❌ An error occurred: {str(e)} \n\n You can open a bug report [here](https://github.com/danielzting/clearurls-discord-bot/issues) 🪳", ephemeral=True)
+        logger.exception(f"Error in linkautomod command: {e}", e)
```
Is {e} necessary in logger.exception?

danielzting commented on main.py:
```diff
@@ -55,38 +75,104 @@ async def on_message(message):
+        logger.info(f"set replace messages setting to {status} for {interaction.guild.id}")
+        
+    except Exception as e:
+        await interaction.response.send_message(f"❌ An error occurred: {str(e)} \n\n You can open a bug report [here](https://github.com/danielzting/clearurls-discord-bot/issues) 🪳", ephemeral=True)
```
Could you change this to not send the error message just in case it contains sensitive information? If you want, you can generate a random UUID, include it in the log message, and give it to the user asking them to attach it to their bug report, but it's not necessary

danielzting replied:
```suggestion
        await interaction.response.send_message(f"❌ An error occurred: {str(e)} \n\n You can open a bug report [on GitHub](https://github.com/danielzting/clearurls-discord-bot/issues) 🪳", ephemeral=True)
```

danielzting commented on main.py:
```diff
@@ -55,38 +75,104 @@ async def on_message(message):
+    
+    reacting_user = await bot.fetch_user(payload.user_id)
+    authorized_user = None
+    
```
Could you write a couple sentences explaining how this get original author logic fits in with the automod setting? Reviewing this logic took a little bit of mental effort even though I already understood what the code was supposed to be doing. Something like we are trying to find the author of the original message to make sure it is they who reacted, and to do that we first try to check the replied-to message, which will be there if automod is off. If automod is on, then it won't exist, so we check the mentions.

But do we even need all this logic—I saw that you mention the author no matter if the automod setting is on or off, so can we just use the "fallback" case directly?

danielzting reviewed (changes requested):
This looks almost good to go, just have some small requests now. Was busy recently and I'm pretty picky with code design/style so appreciate your patience!

teauxfu committed:
lipstick

danielzting commented:
Were you able to test this? I tried running it but got SyntaxErrors on main.py:97

teauxfu commented:
Yes. I'm going to leave it here, feel free to adapt the changes if you like. 

In the future you may consider adding some linters or pre-commit hooks to the project, to reduce friction over style, and so that future contributors can focus on features not capitalization and whitespace when they opt to volunteer their time. 

<img width="215" height="50" alt="image" src="https://github.com/user-attachments/assets/b4ce5286-5959-4a34-953b-458fecb370b6" />
